File: python/fin_ratios/types/financials.py
"""Core financial statement types."""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class IncomeStatement:
    revenue: float
    gross_profit: float
    cogs: float
    ebit: float
    ebitda: float
    net_income: float
    interest_expense: float
    income_tax_expense: float
    ebt: float
    depreciation_and_amortization: float
    shares_outstanding: float
    shares_outstanding_diluted: Optional[float] = None
    eps: Optional[float] = None
    eps_diluted: Optional[float] = None
    research_and_development: Optional[float] = None
    sga_expense: Optional[float] = None
    period: Optional[str] = None


@dataclass
class BalanceSheet:
    # Assets
    total_assets: float
    current_assets: float
    cash: float
    accounts_receivable: float
    inventory: float
    net_ppe: float
    retained_earnings: float
    # Liabilities
    total_liabilities: float
    current_liabilities: float
    accounts_payable: float
    long_term_debt: float
    total_debt: float
    # Equity
    total_equity: float
    # Optional
    short_term_investments: Optional[float] = None
    short_term_debt: Optional[float] = None
    goodwill: Optional[float] = None
    intangible_assets: Optional[float] = None
    shares_outstanding: Optional[float] = None
    period: Optional[str] = None


@dataclass
class CashFlowStatement:
    operating_cash_flow: float
    capex: float
    investing_cash_flow: Optional[float] = None
    financing_cash_flow: Optional[float] = None
    net_change_in_cash: Optional[float] = None
    stock_based_compensation: Optional[float] = None
    dividends_paid: Optional[float] = None
    debt_repayments: Optional[float] = None
    debt_issuance: Optional[float] = None
    depreciation_and_amortization: Optional[float] = None
    period: Optional[str] = None


@dataclass
class FinancialStatements:
    income: IncomeStatement
    balance: BalanceSheet
    cash_flow: CashFlowStatement

File: typescript/src/types/financials.ts
/**
 * Core financial statement types.
 * All monetary values should be in the same currency unit (e.g., USD thousands).
 * Use null for unavailable/inapplicable fields.
 */

export interface IncomeStatement {
  /** Total revenue / net sales */
  revenue: number
  /** Revenue minus cost of goods sold */
  grossProfit: number
  /** Cost of goods sold */
  cogs: number
  /** Earnings before interest and taxes */
  ebit: number
  /** EBIT + Depreciation + Amortization */
  ebitda: number
  /** Net income after all expenses and taxes */
  netIncome: number
  /** Interest expense (positive = expense) */
  interestExpense: number
  /** Income tax expense */
  incomeTaxExpense: number
  /** Earnings before taxes */
  ebt: number
  /** Depreciation and amortization */
  depreciationAndAmortization: number
  /** Research and development expense */
  researchAndDevelopment?: number | null
  /** Selling, general and administrative expense */
  sgaExpense?: number | null
  /** Basic shares outstanding */
  sharesOutstanding: number
  /** Diluted shares outstanding */
  sharesOutstandingDiluted?: number | null
  /** Earnings per share (basic) */
  eps?: number | null
  /** Earnings per share (diluted) */
  epsDiluted?: number | null
  /** Fiscal period label, e.g. "2024-Q1" or "FY2023" */
  period?: string
}

export interface BalanceSheet {
  // Assets
  /** Total assets */
  totalAssets: number
  /** Current assets */
  currentAssets: number
  /** Cash and cash equivalents */
  cash: number
  /** Short-term investments */
  shortTermInvestments?: number | null
  /** Accounts receivable */
  accountsReceivable: number
  /** Inventory */
  inventory: number
  /** Net PP&E (property, plant, equipment) */
  netPPE: number
  /** Goodwill */
  goodwill?: number | null
  /** Intangible assets (excluding goodwill) */
  intangibleAssets?: number | null
  /** Retained earnings */
  retainedEarnings: number

  // Liabilities
  /** Total liabilities */
  totalLiabilities: number
  /** Current liabilities */
  currentLiabilities: number
  /** Accounts payable */
  accountsPayable: number
  /** Short-term debt */
  shortTermDebt?: number | null
  /** Long-term debt */
  longTermDebt: number
  /** Total debt (short-term + long-term) */
  totalDebt: number

  // Equity
  /** Total shareholders' equity */
  totalEquity: number
  /** Number of shares outstanding */
  sharesOutstanding?: number | null

  period?: string
}

export interface CashFlowStatement {
  /** Cash from operating activities */
  operatingCashFlow: number
  /** Capital expenditures (positive = spending, negative also common) */
  capex: number
  /** Cash from investing activities */
  investingCashFlow?: number | null
  /** Cash from financing activities */
  financingCashFlow?: number | null
  /** Net change in cash */
  netChangeInCash?: number | null
  /** Stock-based compensation */
  stockBasedCompensation?: number | null
  /** Dividends paid */
  dividendsPaid?: number | null
  /** Debt repayments */
  debtRepayments?: number | null
  /** Debt issuance */
  debtIssuance?: number | null
  /** Depreciation and amortization (from cash flow statement) */
  depreciationAndAmortization?: number | null

  period?: string
}

/**
 * Convenience bundle of all three statements for a single period.
 */
export interface FinancialStatements {
  income: IncomeStatement
  balance: BalanceSheet
  cashFlow: CashFlowStatement
}
